### PR TITLE
Adding PHIVE to the ddev-webserver image

### LINKED ISSUE #2990: Add Phive to the PHP image?
To install development aids like PHPUnit or similar tools, we use [Phive](https://github.com/phar-io/phive). It would be great to include their Phar in the ddev PHP images, such that it can be used in the same way as composer

### PR DESCRIPTION
**Pull request by NicoHaase**
This PR adds the PHIVE binary to the PHP image used by ddev

---

[PHIVE](https://github.com/phar-io/phive) ("The PHAR Installation and Verification Environment") is a tool to install other PHAR based tools, like PHPUnit, Psalm,... and keep track of them: in a `phive.xml` in your project folder, it can store which version is to be installed, install them as listed in this configuration file, and also update these PHARs if new versions are released. After all, this is like Composer for PHAR files

---

Reopens #3008
Fixes #2990

<a href="https://gitpod.io/#https://github.com/drud/ddev/pull/3009"></a>

### CODE REVIEW

**Comment by NicoHaase** [comment]
So, two findings so far:

Only magerun can be installed using PHIVE, all other tools are not listed there yet. That's only a small improvement over the existing installation, and I don't think that this is worth to be changed.

Additionally, I'm still not able to run and test my change :( As far as I can see, Gitpod is slow to start, and waits endlessly with "Awaiting port 8080". Opening another terminal and running the steps from https://ddev.readthedocs.io/en/stable/developers/building-contributing/ does not work for me, as `ddev version` won't list the tag I've used. So, I hope that PHIVE is usable as I've commited it.

**Comment by rfay** [comment]
I pushed the image as you have it and rebased to upstream/master (so make sure to reset your branch if you work on this)

**Comment by NicoHaase** [comment]
Thanks for pushing the image - I was able to test whether everything works as I would expect it, and that's the case :) 

So, from my side, this looks complete :)

File: cmd/ddev/cmd/global_dotddev_assets/commands/web/phive [hunk 1]
```diff
@@ -0,0 +1,12 @@
+#!/bin/bash
```
Comment (rfay) [inline cmd/ddev/cmd/global_dotddev_assets/commands/web/phive:1]: I'd rather not do this custom/shell command. In general, only things that are heavily used get helper commands like this. The reason is that they clog up `ddev help` and people don't understand what they're for.

File: docs/users/cli-usage.md [hunk 2]
```diff
@@ -14,7 +14,9 @@ Each of these commands has full help. For example, `ddev start -h` or `ddev help
 * `ddev sequelpro`, `ddev sequelace`, and `ddev tableplus` (macOS only, if the app is installed) give access to the Sequel Pro, Sequel Ace, or TablePlus database browser GUIs.
 * `ddev heidisql` (Windows/WSL2 only, if installed) gives access to the HeidiSQL database browser GUI.
 * `ddev import-db` and `ddev export-db` let you import or export a sql or compressed sql file.
-* `ddev composer` lets you run composer (inside the container), for example `ddev composer install` will do a full composer install for you without even needing composer on your computer. See [developer tools](developer-tools.md#ddev-and-composer). Composer version 2 is the default, but you can also configure composer version 1.
+* `ddev composer` lets you run composer (inside the container), for example `ddev composer install` will do a full composer install for you without even needing composer
+  on your computer. See [developer tools](developer-tools.md#ddev-and-composer). Composer version 2 is the default, but you can also configure composer version 1.
```
Comment (rfay) [inline docs/users/cli-usage.md:18]:
```suggestion
* `ddev composer` lets you run composer (inside the container), for example `ddev composer install` will do a full composer install for you without even needing composer on your computer. See [developer tools](developer-tools.md#ddev-and-composer). Composer version 2 is the default, but you can also configure composer version 1.
```

File: docs/users/cli-usage.md [hunk 3]
```diff
@@ -14,7 +14,9 @@ Each of these commands has full help. For example, `ddev start -h` or `ddev help
...
+* `ddev phive` lets you run PHIVE (inside the container), for example `ddev phive install phpunit` will install PHPUnit using PHIVE
```
Comment (rfay) [inline docs/users/cli-usage.md:19]: Remove this, 
```suggestion
```

**Review by rfay — CHANGES_REQUESTED** [review]
Mostly the removal of the custom/shell command here, as I don't think it's warranted.

I don't understand why magerun, magerun2, terminus, platform, and acli can't be installed with phive. Remind me?

File: cmd/ddev/cmd/global_dotddev_assets/commands/web/phive (hunk 1 above)
Comment (NicoHaase) [reply cmd/ddev/cmd/global_dotddev_assets/commands/web/phive]: Done

**Comment by rfay** [comment]
You move FAST!

**Comment by NicoHaase** [comment]
> I don't understand why magerun, magerun2, terminus, platform, and acli can't be installed with phive. Remind me?

As far as I understand PHIVE, `phive list` lists the available packages, and this does not include terminus, platform, nor acli. magerun and magerun2 **are** listed, but to me it looks more consistent if either **all** such tools are installed using PHIVE, or **none**.

**Comment by rfay** [comment]
Oh, you're saying in https://github.com/drud/ddev/pull/3009#issuecomment-843315172 that phive is not a generic phar installer? I would have been far less interested in this if I'd known that.

**Comment by NicoHaase** [comment]
> Oh, you're saying in [#3009 (comment)](https://github.com/drud/ddev/pull/3009#issuecomment-843315172) that phive is not a generic phar installer? I would have been far less interested in this if I'd known that.

That's not what I wanted to say. I've just tried it myself for the first time, and it is possible to use a raw URL to download Terminus. I'll have a try to see whether it works

File: containers/ddev-webserver/Dockerfile [hunk 4]
```diff
@@ -99,11 +99,11 @@ RUN DEBIAN_FRONTEND=noninteractive apt-get -qq install -o Dpkg::Options::="--for
...
-RUN curl -sSL https://github.com/platformsh/platformsh-cli/releases/download/$(curl --silent "https://api.github.com/repos/platformsh/platformsh-cli/releases/latest" | perl -nle'print $& while m{"tag_name": "\K.*?(?=")}g')/platform.phar --output /usr/local/bin/platform && chmod 777 /usr/local/bin/platform
-RUN curl -sSL --output /usr/local/bin/acli https://github.com/acquia/cli/releases/latest/download/acli.phar && chmod 777 /usr/local/bin/acli
-RUN curl -sSL https://github.com/backdrop-contrib/drush/releases/download/${BACKDROP_DRUSH_VERSION}/backdrop-drush-extension.zip -o /tmp/backdrop-drush-extension.zip && unzip -o /tmp/backdrop-drush-extension.zip -d /var/tmp/backdrop_drush_commands
 RUN curl -sSL --output /usr/local/bin/phive "https://phar.io/releases/phive.phar" && chmod 777 /usr/local/bin/phive
+RUN phive install --force-accept-unsigned --target /usr/local/bin/ --temporary pantheon-systems/terminus
```
Comment (rfay) [inline containers/ddev-webserver/Dockerfile:103]: Nice. Now if we could get these signed properly we'd really be somewhere. https://github.com/drud/ddev/issues/2969 is a super-important initiative to improve defenses against supply-chain exploits. Not sure what's required.

I note that there's also a `phive install --global` which might serve the same as your --target /usr/local/bin

Comment (NicoHaase) [reply containers/ddev-webserver/Dockerfile:103]: None of these PHARs are signed :( But the global flag is a good idea

File: containers/ddev-webserver/Dockerfile [hunk 5]
```diff
@@ -100,9 +100,9 @@ ADD ddev-webserver-dev-base-files /
 RUN source /tmp/ddev/vars && curl -sSL https://github.com/drud/MailHog/releases/download/v${MAILHOG_VERSION}/MailHog_${MAILHOG_ARCH} -o /usr/local/bin/mailhog; 
 
 RUN curl -sSL --output /usr/local/bin/phive "https://phar.io/releases/phive.phar" && chmod 777 /usr/local/bin/phive
-RUN phive install --force-accept-unsigned --target /usr/local/bin/ --temporary pantheon-systems/terminus
-RUN phive install --force-accept-unsigned --target /usr/local/bin/ --temporary platformsh/platformsh-cli
-RUN phive install --force-accept-unsigned --target /usr/local/bin/ --temporary acquia/cli
+RUN phive install --force-accept-unsigned --global pantheon-systems/terminus
+RUN phive install --force-accept-unsigned --global platformsh/platformsh-cli
+RUN phive install --force-accept-unsigned --global acquia/cli
```
[the remark this answers is not in the document]
Comment (NicoHaase) [reply containers/ddev-webserver/Dockerfile:105]: Does it really? On my system, this is printed to the shell:

```
Downloading https://api.github.com/rate_limit
Downloading https://github.com/acquia/cli/releases/download/1.12.0/acli.phar
Copying cli-1.12.0.phar to /usr/local/bin/acli
```

File: containers/ddev-webserver/Dockerfile (hunk 4 above)
Comment (NicoHaase) [reply containers/ddev-webserver/Dockerfile]: > I don't understand why magerun, magerun2, terminus, platform, and acli can't be installed with phive. Remind me?

Coming back to magerun and magerun2: they can also be installed using PHIVE, but they would be saved under their "original" names `n98-magerun` and `n98-magerun2`. Adding a symlink to these would be possible, I think

**Comment by rfay** [comment]
I pushed the current image, and added a commit to turn off xdebug during the phive activities. 

I see that acli is there just fine.

**Comment by NicoHaase** [comment]
Just for clarification: for me, everything works fine, and I'm pretty thankful for your help on this branch. Is there anything else from your side that this needs, like more documentation or anything else I've missed?

**Comment by rfay** [comment]
I'm on vacation this week, will be back at it next week. I imagine we can get this in in a couple of weeks, thanks for all the work on it!

**Comment by rfay** [comment]
Rebased and pushed refreshed ddev-webserver

**Comment by rfay** [comment]
Pushed updated image, rebased, quieted output

**Review by rfay — APPROVED** [review]
Pushed latest and rebased. I think when this finishes the automated checks we'll be good to go.

**Comment by rfay** [comment]
Unfortunately, the platform tool installed here gets a SEGV when used. I haven't been able to figure out why, but it's not happening on other currently-update PRs like https://github.com/drud/ddev/pull/, so may be the fault of phive. I opened https://github.com/platformsh/platformsh-cli/issues/1063 but don't know if they'll have any help.

**Comment by NicoHaase** [comment]
> Unfortunately, the platform tool installed here gets a SEGV when used. I haven't been able to figure out why, but it's not happening on other currently-update PRs like [drud/ddev/pull](https://github.com/drud/ddev/pull/), so may be the fault of phive. I opened [platformsh/platformsh-cli#1063](https://github.com/platformsh/platformsh-cli/issues/1063) but don't know if they'll have any help.

When called without a specific version, `phive install` uses the latest release. Have you checked whether manually requiring a version helps? Or does the exact same version of platform fail when being installed using phive, but not when installed the old way?

**Comment by rfay** [comment]
The weird thing is that I can get the SEGV with any version, manually installed. So it's possible it's something wrong on the server side with Platform.

**Comment by rfay** [comment]
Somehow phive is installing platform 4.1.0-dev

> rfay@d9-web:/var/www/html$ platform --version
Platform.sh CLI 4.1.0-dev

**Comment by rfay** [comment]
Removing the phive-based installation of other tools solved the build problem. 

It's probably important to figure out what phive installs. It doesn't seem to install the latest release, which would be what most people would want. 

How much have you used it @NicoHaase ? It seems like it may have unwanted behaviors.  It also doesn't seem like a project that has that much traction in the php world. How did you run across it?

**Comment by NicoHaase** [comment]
I've opened an issue at their bugtracker to ask for clarification (see the link above), and to me this looks like a problem with Phive itself.

The "traction in the PHP world" might be discussable. Yes, that is definitely not a tool as common as Composer, but it is backed and recommended by people like Sebastian Bergmann (the creator of PHPUnit)

**Comment by rfay** [comment]
I sure like the idea of automated checksum checking.

I don't see the bugtrack link you mentioned "above", please include it in your comment, thanks.

**Comment by NicoHaase** [comment]
Sorry, https://github.com/phar-io/phive/issues/313 is the bug report I was talking about

**Review by rfay — APPROVED** [review]
Thanks for all your work on this. I'm in the process of pushing a rebased image and rebasing the base code, and then when this gets through tests it will be good for merge. Sorry the effort of using phive to install various phars didn't work out here, but I hope phive has a bright future. I do hope they'll focus on released versions.

**Comment by rfay** [comment]
Thanks for all the good work on this @NicoHaase , and congrats on first PR in here, hoping it's the first of many.

**Pull request merged**